### PR TITLE
[Core] Update json library and improve Parameters

### PR DESCRIPTION
**Pull request by loumalouomega**
This updates the version of the library to 3.7.3. The main difference is the items object which allows to simplify the code (and will allow even more simplification if we adopt C++17). Additionally I add the SetTYPE methods in order to simplify the modification of the Parameters (adding new values is actually a little bit painful), this simplifies that. I also added the Write Read methods in order to read or write json files

### CODE REVIEW

**Comment by loumalouomega** [comment]
I also added the Write Read methods in order to read or write json files

**Comment by adityaghantasala** [comment]
I would suggest to add a new constructor to KratosParameters which accepts the JSON file name and uses the new ReadJSONFile function.

**Comment by loumalouomega** [comment]
> I would suggest to add a new constructor to KratosParameters which accepts the JSON file name and uses the new ReadJSONFile function.

:+1: (the problem is that the std::string constructor already exists, and I don't know how to solve that problem)

**Comment by adityaghantasala** [comment]
What is the problem with constructor with String as argument ?

**Comment by loumalouomega** [comment]
> What is the problem with constructor with String as argument ?

It already exists, so some check must be done...

**Comment by philbucher** [comment]
In the future this could be solved with a constructor that accepts an std::filesystem::path (another point for C++17 :) )

**Comment by loumalouomega** [comment]
Indeed

File: kratos/sources/kratos_parameters.cpp [hunk 1]
```diff
@@ -687,6 +714,84 @@ void Parameters::SetMatrix(const Matrix& rValue)
 /***********************************************************************************/
 /***********************************************************************************/
 
+void Parameters::AddDouble(
```
Comment (GuillermoCasas) [inline kratos/sources/kratos_parameters.cpp:700]: Nice utilities!

**Comment by GuillermoCasas** [comment]
Thank you! It looks alright to me.

**Comment by AlejandroCornejo** [comment]
O behalf of @KratosMultiphysics/implementation-committee, We agree with this implementation. Feel free to merge from our side. Now the @KratosMultiphysics/technical-committee must approve.

**Comment by roigcarlo** [comment]
@KratosMultiphysics/technical-committee we have some suggestions:
- AddXXX methods could be templated
- Instead of ReadJSONFile, please create a constructor templated in the stream type (see: https://stackoverflow.com/questions/33628250/c-reading-a-json-object-from-file-with-nlohmann-json)
- Remove the WriteJSONFile, as there are methods to convert parameters to string, which can then be used to write to a file

Regarding 2 & 3, this is due to the fact that we will handle json includes using streams.

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/technical-committee we have some suggestions:
> 
>     * AddXXX methods could be templated
Is not possible, for the same reason SetXXX cannot be.
> 
>     * Instead of ReadJSONFile, please create a constructor templated in the stream type (see: https://stackoverflow.com/questions/33628250/c-reading-a-json-object-from-file-with-nlohmann-json)
> 
OK
>     * Remove the WriteJSONFile, as there are methods to convert parameters to string, which can then be used to write to a file
OK
> 
> 
> Regarding 2 & 3, this is due to the fact that we will handle json includes using streams.

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/technical-committee we have some suggestions:
> 
>     * AddXXX methods could be templated
> 
>     * Instead of ReadJSONFile, please create a constructor templated in the stream type (see: https://stackoverflow.com/questions/33628250/c-reading-a-json-object-from-file-with-nlohmann-json)
> 
>     * Remove the WriteJSONFile, as there are methods to convert parameters to string, which can then be used to write to a file
> 
> 
> Regarding 2 & 3, this is due to the fact that we will handle json includes using streams.

I applied your suggestions

**Comment by loumalouomega** [comment]
This is ready

**Comment by loumalouomega** [comment]
Ping

**Review by roigcarlo — APPROVED** [review]
:+1:

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee approves :+1:

**Comment by loumalouomega** [comment]
Hurray

**Comment by loumalouomega** [comment]
Ironically today 3.9 was released: https://github.com/nlohmann/json/releases/tag/v3.9.0, this is an important release which have an important improvement: **It allows comments using /* */  or // ala C**, this may be very interesting to document examples and tests, opinion @KratosMultiphysics/technical-committee ?

**Comment by philbucher** [comment]
We saw that too and were wondering when you make the PR :D

To me makes sense to use it, eventually we would update anyway
Any having comments I guess is not a bad thing (my opinion)

**Pull request merged**